queries/java/locals.scm

```diff
@@ -1,10 +1,104 @@
-; CREDITS @maxbrunsfeld (maxbrunsfeld@gmail.com)
-
+; SCOPES
+; declarations
+(program) @scope
 (class_declaration
-  name: (identifier) @name) @class
+  body: (_) @scope)
+(enum_declaration
+  body: (_) @scope)
+(method_declaration) @scope ; whole method_declaration because arguments
+
+; block
+(block) @scope
+
+; if/else
+(if_statement) @scope ; if+else
+(if_statement
+  consequence: (_) @scope) ; if body in case there are no braces
+(if_statement
+  alternative: (_) @scope) ; else body in case there are no braces
+
+; try/catch
+(try_statement) @scope ; covers try+catch, individual try and catch are covered by (block)
+(catch_clause) @scope ; needed because `Exception` variable
+
+; loops
+(for_statement) @scope ; whole for_statement because loop iterator variable
+(for_statement         ; "for" body in case there are no braces
+  body: (_) @scope)
+(do_statement
+  body: (_) @scope)
+(while_statement
+  body: (_) @scope)
+
+; Functions
+
+(constructor_declaration) @scope
+(method_declaration
+  name: (identifier) @definition.scope) @scope
+
 
+; DEFINITIONS
+ (package_declaration
+    (identifier) @definition.namespace) 
+(class_declaration
+  name: (identifier) @definition.class)
+(enum_declaration
+  name: (identifier) @definition.enum)
 (method_declaration
-  name: (identifier) @name) @method
+  name: (identifier) @definition.method)
+
+(local_variable_declaration
+  declarator: (variable_declarator
+                name: (identifier) @definition.var))
+(formal_parameter
+  name: (identifier) @definition.var)
+(catch_formal_parameter
+  name: (identifier) @definition.var)
+(inferred_parameters (identifier) @definition.var) ; (x,y) -> ...
+(lambda_expression
+    parameters: (identifier) @definition.var) ; x -> ...
+
+; we need submatch!
+(import_declaration
+ (scoped_identifier
+   ((identifier) @definition.import)))
+(import_declaration
+  (scoped_identifier
+    (scoped_identifier
+      ((identifier) @definition.import))))
+(import_declaration
+  (scoped_identifier
+    (scoped_identifier
+      (scoped_identifier
+        ((identifier) @definition.import)))))
+(import_declaration
+  (scoped_identifier
+    (scoped_identifier
+      (scoped_identifier
+        (scoped_identifier
+          ((identifier) @definition.import))))))
+(import_declaration
+  (scoped_identifier
+    (scoped_identifier
+      (scoped_identifier
+        (scoped_identifier
+          (scoped_identifier
+            ((identifier) @definition.import)))))))
+(import_declaration
+  (scoped_identifier
+    (scoped_identifier
+      (scoped_identifier
+        (scoped_identifier
+          (scoped_identifier
+            (scoped_identifier
+              ((identifier) @definition.import))))))))
```

Comment:
We really need submatch, can't we use something like `(scoped_identifier (identifier))` alone ?

Reply:
This would also match usages of the classes like `google.FooClass()`. I can finish the predicates PR today with a simplified version of submatch (only looks for a specific node type as parent).

A real submatch is difficult to implement with current query syntax.

Reply:
No rush, maybe we should ask the question in upstream treesitter.

Reply:
Something like a new wildcard, or a new meaning for `_`.

Reply:
I removed this for now. Can be solved e.g. with the predicates PR #208